### ISSUE TITLE
tradeEnv Issue

### ISSUE THREAD

**Issue by lrhgogogo** [SteamerLee/MASA#8]
![1729677549556](https://github.com/user-attachments/assets/385f2df8-4363-4b22-8609-d9cfd8cadb88)
May I ask a question? In the end of save_profile() of tradeEnv.py, why we need to reset step_data when (invest_profile['ep'] == self.config.num_epochs), meaning that when it comes to final episode like 50, step_data will always be the detail info of episode_50 and then be saved. Maybe we should save step_data in the IF instead of reset it? I not sure, did I misunderstand something?

**Comment by lrhgogogo** [comment]
I got it, the columns saved each time are different, excellent!!!

**Comment by lrhgogogo** [comment]
![1729690917520](https://github.com/user-attachments/assets/65547174-4f1c-4ba2-bf6a-c043b6329a3e)
![1729690949788](https://github.com/user-attachments/assets/24fd8de4-ab4c-4e7f-92e0-10b35ba6ea24)
Its weird that in the final valid_profile and test_profile, adjacent two episode got the same result after a certain episode(3 in my situation DJIA-MASA-DC). This situation still occurs when I re-executed entrance.py and I haven't found the reason yet. By the way is it convinient for you to chat in Chinese hah. Thank you for you kind reply!!

**Comment by SteamerLee** [comment]
> I got it, the columns saved each time are different, excellent!!!

Good! But please be aware that those results are just for analyzing the model. When reporting the results, it should use the best model validated in the validation set to evaluate and report the performance in the test set. (e.g., 'xxx_policy_validbest' columns in the test_stepdata.csv, or the results in ep='x' in the test_profile.csv and 'x' comes from the ep in the valid_bestmodel.csv.)

Thank you!

**Comment by SteamerLee** [comment]
> ![1729690917520](https://private-user-images.githubusercontent.com/130480199/379295679-65547174-4f1c-4ba2-bf6a-c043b6329a3e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3MDYwODQsIm5iZiI6MTcyOTcwNTc4NCwicGF0aCI6Ii8xMzA0ODAxOTkvMzc5Mjk1Njc5LTY1NTQ3MTc0LTRmMWMtNGJhMi1iZjZhLWMwNDNiNjMyOWEzZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyM1QxNzQ5NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZDNkMDk3NjM2ZWJhNTg3MDU1ZjI5ZjRiNTcxMTUzYjZmZDMwOTlhOTA4ZTY5ZTM4OGQzMjVhZDQyNWRlYWZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.S_3ZAsMkSQk0Be3WGSSUoT5jJcQ5lzofmqOlNbJg36c) ![1729690949788](https://private-user-images.githubusercontent.com/130480199/379295868-24fd8de4-ab4c-4e7f-92e0-10b35ba6ea24.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3MDYwODQsIm5iZiI6MTcyOTcwNTc4NCwicGF0aCI6Ii8xMzA0ODAxOTkvMzc5Mjk1ODY4LTI0ZmQ4ZGU0LWFiNGMtNGU3Zi05MmUwLTEwYjM1YmE2ZWEyNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyM1QxNzQ5NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNDllYTAyMDkwZmM0ZmEwNWI2ZTcwNWI1N2E3ODUwMDhkMjMyN2JjNDZmYmQ2ODBjYTA4Njc3OGZhYjhjOWNlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.iJEekBSgwGJQYiY0_p2O5142vs83tTpKGJi2vuXi_VQ) Its weird that in the final valid_profile and test_profile, adjacent two episode got the same result after a certain episode(3 in my situation DJIA-MASA-DC). This situation still occurs when I re-executed entrance.py and I haven't found the reason yet. By the way is it convinient for you to chat in Chinese hah. Thank you for you kind reply!!

It may come from the settings of TD3 in the stable-baseline3. The update frequency of the action network is set to 2 epoches such that you may observe the results updated by the same frequency. 

Sure, both Chinese and English work for me. But I may not regularly check github and immediately reply questions since the manpower is limit and I am super busy in these days.

Thank you for your kind attention again!

**Comment by lrhgogogo** [comment]
由衷地感谢，祝你生活愉快！！！

**Comment by lrhgogogo** [comment]
![1729848766192](https://github.com/user-attachments/assets/99bb421a-2523-4521-93dc-959c45f3f3f2)
![1729849204480](https://github.com/user-attachments/assets/0907fe30-ec7d-402b-824f-f499ff4d646f)

我发现了一个小问题：在tradeEnv.py的save_profile()中（上图），假设X_best_epi表示在当前数据集中表现最佳的epi号，目前的逻辑当valid_best_epi < test_best_epi时，会导致最后保存的test_stepdata中"best"相关的列信息对应的是test集下最佳的epi，而不是valid集下的
因此我进行了简单的修改，或许能够帮助到这一小问题（下图）

另外我的训练过程也有一个奇怪的现象：三种市场下，三种模型（masa-dc,masa-mlp,masa-lstm）中，valid_best_epi经常是2或者49，即模型要么早早地就达到了“最优”或是即使epi完全结束也可能没有达到“最优”，这与我的数据源有关吗（目前没有使用60m级数据，仅使用了1d数据），我还在思考这个问题。

Please feel free to respond at your convenience. I truly appreciate your valuable time. Thank you!

**Comment by SteamerLee** [comment]
您好，关于第一个问题，test_stepdata的'xxx_best'中储存的是test集下的最优ep，这个只是用于分析性能。而best validated model在test set下的结果是在'xxx_validbest'中，您可以直接查看该列以获得想要的观察数据。

关于第二个问题，这与数据，模型训练等因素有关系，也与金融数据的特殊性质有关，这都可能增加训练难度。您可以尝试增加epoch，更换不同kernel model等去观察。

谢谢！
